state/watcher/hubwatcher.go

```diff
@@ -185,6 +213,81 @@ func (w *HubWatcher) UnwatchCollection(collection string, ch chan<- Change) {
 	w.sendReq(reqUnwatch{watchKey{collection, nil}, ch})
 }
 
+// HubWatcherStats defines a few metrics that the hub watcher tracks
+type HubWatcherStats struct {
+	// WatchKeyCount is the number of keys being watched
+	WatchKeyCount int
+	// WatchCount is the number of watchers (keys can be watched by multiples)
+	WatchCount uint64
+	// RevnoMapSize is the number of keys stored in the 'current' map
+	RevnoMapSize int
+	// RevnoMapBytes is an approximation of the number of bytes being stored by 'current'
+	RevnoMapBytes uint64
+	// SyncQueueCap is the maximum buffer size for synchronization events
+	SyncQueueCap int
+	// SyncQueueLen is the current number of events being queued
+	SyncQueueLen int
+	// SyncLastLen was the length of SyncQueue the last time we flushed
+	SyncLastLen int
+	// SyncAvgLen is a smoothed average of recent sync lengths
+	SyncAvgLen int
+	// SyncMaxLen was the longest we've seen SyncQueue when flushing
+	SyncMaxLen int
+	// SyncEventDocCount is the number of sync events we've generated for specific documents
+	SyncEventDocCount uint64
+	// SyncEventDocCount is the number of sync events we've generated for documents changed in collections
+	SyncEventCollCount uint64
+	// RequestEventCount is the number of request events we've generated
+	// (documents being watched that changed since the request came in)
+	RequestEventCount uint64
+	// RequestQueueCap is the maximum size of the request queue buffer
+	RequestQueueCap int
+	// RequestQueueCap is the current number of requested events
+	RequestQueueLen int
+	// RequestCount is the number of requests (reqWatch/reqUnwatch, etc) that we've seen
+	RequestCount uint64
+	// ChangeCount is the number of changes we've processed
+	ChangeCount uint64
+}
+
+type reqStats struct {
+	ch chan<- HubWatcherStats
+}
+
+func (w *HubWatcher) Stats() HubWatcherStats {
+	ch := make(chan HubWatcherStats, 1)
```

Comment:
So not buffering it makes the loop() block until we do the <- ch. Which is ok in the sense that we're just about to do that anyway. But it didn't feel nice to me that we would send a blocking channel into loop().
I'll get rid of it since you clearly don't prefer buffered channels.